sdk/clientcore/core/src/main/java/io/clientcore/core/http/models/SdkRequestContext.java

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.core.http.models;
+
+import io.clientcore.core.instrumentation.InstrumentationContext;
+import io.clientcore.core.instrumentation.logging.ClientLogger;
+import io.clientcore.core.utils.ProgressReporter;
+
+import java.util.function.Consumer;
+
+/**
+ * This class contains the request metadata that flows along with the {@link HttpRequest} as it goes through HTTP pipeline.
+ * It includes {@link RequestOptions} provided by the user along with metadata provided by client libraries and core components.
+ * <p>
+ * This class is intended to be used by client library code and HTTP pipeline components. Application developers should
+ * configure request metadata using {@link RequestOptions}.
+ * <p>
+ * This class does not allow any modifications to the request metadata provided by the user.
+ */
+public final class SdkRequestContext extends RequestOptions {
```

Comment:
add Fluent and Immutable, need to find reliable way to ensure it stays immutable